lib/puppet/util/windows/service.rb

```diff
@@ -0,0 +1,726 @@
+        # value of pcbBytesNeeded
+        FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+          # return value will be false from this call, since it's designed
+          # to fail. Just ignore it
+          QueryServiceConfigW(service, nil, 0, bytes_pointer)
+          size_required = bytes_pointer.read(:dword)
+        end
+        FFI::MemoryPointer.new(size_required) do |ssp_ptr|
+          config = QUERY_SERVICE_CONFIGW.new(ssp_ptr)
+          FFI::MemoryPointer.new(:lpword) do |bytes_pointer|
+            if QueryServiceConfigW(
+              service,
+              ssp_ptr,
+              size_required,
+              bytes_pointer
+            ) == FFI::WIN32_FALSE
+              raise Puppet::Util::Windows::Error.new(_("Service query failed"))
+            end
+          end
+        end
+        config
+      end
+      private :query_config
+
+      # @api private
+      # waits for a windows service to report final_state if it
+      # is in pending_state
+      #
+      # @param [:handle] service handle to the service to wait on
+      # @param [Integer] pending_state the state to wait on
+      # @param [Integer] final_state the state indicating the transition is finished
+      # @return [bool] 'true' once the service is reporting final_state,
+      #   'false' if the service was not in pending_state or finaL_state
+      def wait_for_pending_transition(service, pending_state, final_state)
+        elapsed_time = 0
+        last_checkpoint = -1
+        loop do
+          status = query_status(service)
+          state = status[:dwCurrentState]
+          return true if state == final_state
+          unless state == pending_state
+            return false
+          end
+          # When the service is in the pending state we need to do the following:
+          # 1. check if any progress has been made since dwWaitHint using dwCheckPoint,
+          #    and fail if no progress was made
+          # 2. if progress has been made, increment elapsed_time and set last_checkpoint
+          # 3. sleep, then loop again if there was progress.
+          time_to_wait = process_wait_hint(status[:dwWaitHint])
+          if status[:dwCheckPoint] > last_checkpoint
+            elapsed_time = 0
+          else
+            if elapsed_time > (status[:dwWaitHint] / 1000)
+              raise Puppet::Error.new(_("No progress made on service operation and dwWaitHint exceeded"))
+            end
+            elapsed_time += time_to_wait
+          end
+          last_checkpoint = status[:dwCheckPoint]
+          sleep(time_to_wait)
+        end
+      end
+      private :wait_for_pending_transition
+
+      # @api private
+      #
+      # process the wait hint listed by a service to something
+      # usable by ruby sleep
+      #
+      # @param [Integer] wait_hint the wait hint of a service in milliseconds
+      # @return [Integer] the time to wait in seconds between querying the service
+      def process_wait_hint(wait_hint)
```

Comment:
Might be better to rename this `wait_hint_to_wait_time`. `process_wait_hint` makes it seem like there's way more stuff going on in this method.